### PR TITLE
RFC 7539 - ChaCha20 and Poly1305

### PR DESCRIPTION
**Pull request by karel-m**
Basically the reference implementation adopted to our C macros and our calling conventions.

https://tools.ietf.org/html/rfc7539

### CODE REVIEW

**Comment by sjaeckel** [comment]
1. as it's a stream cipher, would it make sense to integrate it in the prng environment as all the other stream ciphers?

    or should we think about integrating resp. how to integrate the stream ciphers in the crypt environment?

    or should we provide an alternative environment or re-think the entire approach?

2. what do you think about this?

In my opinion we shouldn't introduce new algorithms with `unsigned char*` for in- or output data.

```
diff --git a/src/ciphers/chacha/chacha_crypt.c b/src/ciphers/chacha/chacha_crypt.c
index 1ac3aa7..359e485 100644
--- a/src/ciphers/chacha/chacha_crypt.c
+++ b/src/ciphers/chacha/chacha_crypt.c
@@ -51,10 +51,11 @@ static void _chacha_block(unsigned char *output, const ulong32 *input, int round
    @param out     [out] The ciphertext (or plaintext), length inlen
    @return CRYPT_OK if successful
 */
-int chacha_crypt(chacha_state *st, const unsigned char *in, unsigned long inlen, unsigned char *out)
+int chacha_crypt(chacha_state *st, const void *pin, unsigned long inlen, unsigned char *out)
 {
    unsigned char buf[64];
    unsigned long i, j;
+   const unsigned char *in = pin;
 
    if (inlen == 0) return CRYPT_OK;
    if (st->ksleft > 0) {
diff --git a/src/ciphers/chacha/chacha_test.c b/src/ciphers/chacha/chacha_test.c
index d678e93..7218425 100644
--- a/src/ciphers/chacha/chacha_test.c
+++ b/src/ciphers/chacha/chacha_test.c
@@ -41,16 +41,16 @@ int chacha_test(void)
    /* crypt piece by piece */
    chacha_setup(&st, k, sizeof(k), 20);
    chacha_ivctr32(&st, n, sizeof(n), 1);
-   chacha_crypt(&st, (unsigned char*)pt,      35,       out);
-   chacha_crypt(&st, (unsigned char*)pt + 35, 35,       out + 35);
-   chacha_crypt(&st, (unsigned char*)pt + 70,  5,       out + 70);
-   chacha_crypt(&st, (unsigned char*)pt + 75,  5,       out + 75);
-   chacha_crypt(&st, (unsigned char*)pt + 80, len - 80, out + 80);
+   chacha_crypt(&st, pt,      35,       out);
+   chacha_crypt(&st, pt + 35, 35,       out + 35);
+   chacha_crypt(&st, pt + 70,  5,       out + 70);
+   chacha_crypt(&st, pt + 75,  5,       out + 75);
+   chacha_crypt(&st, pt + 80, len - 80, out + 80);
    if (compare_testvector(out, len, ct, sizeof(ct), "CHACHA-TV1", 1) != 0) return CRYPT_FAIL_TESTVECTOR;
    /* crypt in one go */
    chacha_setup(&st, k, sizeof(k), 20);
    chacha_ivctr32(&st, n, sizeof(n), 1);
-   chacha_crypt(&st, (unsigned char*)pt, len, out);
+   chacha_crypt(&st, pt, len, out);
    if (compare_testvector(out, len, ct, sizeof(ct), "CHACHA-TV2", 1) != 0) return CRYPT_FAIL_TESTVECTOR;
    /* crypt in one go - using chacha_ivctr64() */
    chacha_setup(&st, k, sizeof(k), 20);
diff --git a/src/headers/tomcrypt_cipher.h b/src/headers/tomcrypt_cipher.h
index f26115d..e38d680 100644
--- a/src/headers/tomcrypt_cipher.h
+++ b/src/headers/tomcrypt_cipher.h
@@ -819,7 +819,7 @@ typedef struct {
 int chacha_setup(chacha_state *st, const unsigned char *key, unsigned long keylen, int rounds);
 int chacha_ivctr32(chacha_state *st, const unsigned char *iv, unsigned long ivlen, ulong32 counter);
 int chacha_ivctr64(chacha_state *st, const unsigned char *iv, unsigned long ivlen, ulong64 counter);
-int chacha_crypt(chacha_state *st, const unsigned char *in, unsigned long inlen, unsigned char *out);
+int chacha_crypt(chacha_state *st, const void *pin, unsigned long inlen, unsigned char *out);
 int chacha_keystream(chacha_state *st, unsigned char *out, unsigned long *outlen);
 int chacha_test(void);
 #endif /* LTC_CHACHA */
```

File: src/headers/tomcrypt_custom.h [hunk 1]
```diff
@@ -262,6 +265,7 @@
 #define LTC_OCB3_MODE
 #define LTC_CCM_MODE
 #define LTC_GCM_MODE
+#define LTC_CHACHA20POLY1305_MODE
```
Comment (sjaeckel) [inline src/headers/tomcrypt_custom.h]: I guess we should add a sanity check in the cleanup section

```
#if defined(LTC_CHACHA20POLY1305_MODE) && !defined(LTC_POLY1305) ...
```

File: src/encauth/chachapoly/chacha20poly1305_done.c [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+int chacha20poly1305_done(chachapoly_state *st, unsigned char *tag, unsigned long *taglen)
+{
+   unsigned char padzero[16] = { 0 };
+   unsigned long padlen;
+   unsigned char buf[16];
+   int err;
+
+   if (*taglen != 16) return CRYPT_ERROR;
+   if ((padlen = 16 - (st->ctlen % 16)) < 16) {
```
Comment (sjaeckel) [inline src/encauth/chachapoly/chacha20poly1305_done.c]: assignment please in a separate line

**Comment by karel-m** [comment]
Ad `unsigned char*` vs. `void *` I am not against that change but I prefer to make a BIG SWITCH and change all such cases in the whole libtomcrypt. Not just in a single function.

File: src/headers/tomcrypt_custom.h (hunk 1 above)
Comment (karel-m) [inline src/headers/tomcrypt_custom.h]: I agree

File: src/encauth/chachapoly/chacha20poly1305_done.c (hunk 2 above)
Comment (karel-m) [inline src/encauth/chachapoly/chacha20poly1305_done.c]: I agree

**Comment by karel-m** [comment]
Ad moving `chacha` implementation under `prng`: 
It would be probably logical to have all stream cipher in one place, however I am not a fan of the idea of having stream ciphers in `prng` in general. It is IMO very confusing for newcomers and frankly speaking if you want to use RC4 or SOBER128 for encrypting/decrypting your first impression might be that libtomcrypt is not supporting it (users will be looking for a simple `rc4_setup(key)` > `rc4_crypt(in, out)` > `rc4_crypt(in, out)` > `...`  interface which we do not have).

We can create `src/stream` and I can move `src/cipher/chacha/*` to `src/stream/chacha/*`

What I would like more than current state of RC4 + SOBER128
* having `src/stream/rc4.c` and * having `src/stream/sober128.c` which will provide cipher_setup/cipher_crypt interface
* and turning `src/prngs/rc4.c` and `src/prngs/sober128.c` into thin wrappers around the functions from previous point

**Comment by karel-m** [comment]
Let's move `size_t` discussion to  #175 and `unsigned char*` vs. `void *` to #174 so that they do not block this PR.

**Comment by karel-m** [comment]
We have the following:
```C
/* by definition of poly1305 maclen is always 16 */
int poly1305_done(poly_state *st, unsigned char *mac, unsigned long *maclen);
int poly1305_memory(const unsigned char *key, unsigned long keylen, const unsigned char *in, unsigned long inlen, unsigned char *mac, unsigned long *maclen);
int poly1305_memory_multi(const unsigned char *key, unsigned long keylen, unsigned char *mac, unsigned long *maclen, const unsigned char *in,  unsigned long inlen, ...);
int poly1305_file(const char *fname, const unsigned char *key, unsigned long keylen, unsigned char *mac, unsigned long *maclen);
```

What's the best practice in these cases?
1. `int poly1305_done(poly_state *st, unsigned char *mac, unsigned long *maclen)` ?
1. `int poly1305_done(poly_state *st, unsigned char *mac, unsigned long maclen)` ?
1. `int poly1305_done(poly_state *st, unsigned char *mac)` ?
1. `int poly1305_done(poly_state *st, unsigned char mac[16])` ?

**Comment by sjaeckel** [comment]
I'd say 1.

`int poly1305_done(poly_state *st, unsigned char *mac, unsigned long *maclen)`

also I'd prefer 
1. to use the established error codes `CRYPT_INVALID_ARG` for invalid buffer sizes like the `maclen` and `CRYPT_INVALID_KEYSIZE` in e.g. `chacha_setup()` et. al.

2. to be less strict on these buffer sizes

so in `poly1305_done()` it'd become

```
if (*maclen < 16) return CRYPT_INVALID_ARG;
...
*maclen = 16;
return...
```

instead of 

`if (*maclen != 16) return CRYPT_ERROR;`

**Comment by karel-m** [comment]
Ad `unsigned long outlen` vs. `unsigned long *outlen` there 3 more places I have found where we are not consistent (I'll create a separate issue)

**Comment by buggywhip** [comment]
> On Mar 22, 2017, at 5:13 PM, karel-m <notifications@github.com> wrote:
> Ad moving chacha implementation under prng:
> It would be probably logical to have all stream cipher in one place, however I am not a fan of the idea of having stream ciphers in prng in general.
> 
Agreed.  IMO all stream ciphers would be better served under src/stream/, not under prng.  

Many algorithms can be used in ways to produce output normally the province of other classes of algorithms, but without a formal ontology, a "typical"/"traditional" organization structure is probably best.  

And it can get really messy trying to classify authenticated encryption.  ...a unique classification by itself?  ...a mode?  Arrgghh.  :-/

**Comment by karel-m** [comment]
Rebased & squashed into 1 commit, travis-ci all green. I propose merging to develop.

**Comment by karel-m** [comment]
@sjaeckel do you (dis)agree with merging this one?

File: src/prngs/chacha.c [hunk 3]
```diff
@@ -0,0 +1,213 @@
+/* LibTomCrypt, modular cryptographic library -- Tom St Denis
```
Comment (sjaeckel) [inline src/prngs/chacha.c]: Did you follow some guide "how to use ChaCha as a prng" or is this all your work?

From my point of view it looks very good, but if you created this I'd like to have someone else with crypto expertise to review this code.

**Comment by karel-m** [comment]
AFAIK the only unusual magic here is in `chacha_prng_add_entropy`, either initially (before make_ready) and later (kind of rekey operation). 

The implementation (namely the rekey operation) is based on chacha-based arc4random implemented in OpenBSD/FreeBSD (I do not have a link handy but you can easily google it).

**Comment by karel-m** [comment]
Seeing:
```
diff --git a/src/ciphers/chacha/chacha_test.c b/src/ciphers/chacha/chacha_test.c
```
I am in doubts whether you have pulled (you might need `pull --force`) the latest version as I have moved `src/ciphers/chacha` to `src/stream/chacha`.

**Comment by sjaeckel** [comment]
> I am in doubts whether you have pulled (you might need pull --force) the latest version as I have moved src/ciphers/chacha to src/stream/chacha.

simply `s/ciphers/stream/g` in the diff and applying the patch is successful

File: src/stream/chacha/chacha_test.c [hunk 4]
```diff
@@ -24,17 +24,25 @@ int chacha_test(void)
...
+   unsigned char n[]   = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x4a, 0x00, 0x00, 0x00, 0x00 };
+   unsigned char ct[]  = { 0x6E, 0x2E, 0x35, 0x9A, 0x25, 0x68, 0xF9, 0x80, 0x41, 0xBA, 0x07, 0x28, 0xDD, 0x0D, 0x69, 0x81,
+                           0xE9, 0x7E, 0x7A, 0xEC, 0x1D, 0x43, 0x60, 0xC2, 0x0A, 0x27, 0xAF, 0xCC, 0xFD, 0x9F, 0xAE, 0x0B,
+                           0xF9, 0x1B, 0x65, 0xC5, 0x52, 0x47, 0x33, 0xAB, 0x8F, 0x59, 0x3D, 0xAB, 0xCD, 0x62, 0xB3, 0x57,
+                           0x16, 0x39, 0xD6, 0x24, 0xE6, 0x51, 0x52, 0xAB, 0x8F, 0x53, 0x0C, 0x35, 0x9F, 0x08, 0x61, 0xD8,
+                           0x07, 0xCA, 0x0D, 0xBF, 0x50, 0x0D, 0x6A, 0x61, 0x56, 0xA3, 0x8E, 0x08, 0x8A, 0x22, 0xB6, 0x5E,
+                           0x52, 0xBC, 0x51, 0x4D, 0x16, 0xCC, 0xF8, 0x06, 0x81, 0x8C, 0xE9, 0x1A, 0xB7, 0x79, 0x37, 0x36,
+                           0x5A, 0xF9, 0x0B, 0xBF, 0x74, 0xA3, 0x5B, 0xE6, 0xB4, 0x0B, 0x8E, 0xED, 0xF2, 0x78, 0x5E, 0x42,
+                           0x87, 0x4D };
+   unsigned char ct2[] = { 0x6E, 0x2E, 0x35, 0x9A, 0x25, 0x68, 0xF9, 0x80, 0x41, 0xBA, 0x07, 0x28, 0xDD, 0x0D, 0x69, 0x81,
```
Comment (sjaeckel) [inline src/stream/chacha/chacha_test.c]: why should we duplicate the same test-result?

File: src/encauth/chachapoly/chacha20poly1305_decrypt.c [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+{
+   unsigned char padzero[16] = { 0 };
+   unsigned long padlen;
+   int err;
+
+   if (inlen == 0) return CRYPT_OK; /* nothing to do */
+   LTC_ARGCHK(st != NULL);
+
+   if (st->aadflg) {
+      if ((padlen = 16 - (st->aadlen % 16)) < 16) {
```
Comment (karel-m) [inline src/encauth/chachapoly/chacha20poly1305_decrypt.c]: done

File: src/encauth/chachapoly/chacha20poly1305_setiv_rfc7905.c [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+  Set IV + counter data (with RFC7905-magic) to the ChaCha20Poly1305 state and reset the context
+  @param st     The ChaCha20Poly1305 state
+  @param iv     The IV data to add
+  @param inlen  The length of the IV (must be 12 or 8)
+  @param sequence_number   64bit sequence number which is incorporated into IV as described in RFC7905
+  @return CRYPT_OK on success
+ */
+int chacha20poly1305_setiv_rfc7905(chachapoly_state *st, const unsigned char *iv, unsigned long ivlen, ulong64 sequence_number)
```
Comment (karel-m) [inline src/encauth/chachapoly/chacha20poly1305_setiv_rfc7905.c]: and perhaps also `chacha20poly1305_memory`

Comment (sjaeckel) [inline src/encauth/chachapoly/chacha20poly1305_setiv_rfc7905.c]: If you're upto it, sure :+1:

Comment (karel-m) [inline src/encauth/chachapoly/chacha20poly1305_setiv_rfc7905.c]: done

File: src/headers/tomcrypt_mac.h [hunk 7]
```diff
@@ -96,6 +96,26 @@ void pmac_shift_xor(pmac_state *pmac);
...
 
+#ifdef LTC_POLY1305
+typedef struct {
+   ulong32 r[5];
+   ulong32 h[5];
+   ulong32 pad[4];
+   unsigned long leftover;
+   unsigned char buffer[16];
+   int final;
+} poly_state;
```
Comment (karel-m) [inline src/headers/tomcrypt_mac.h]: done

File: src/headers/tomcrypt_mac.h [hunk 8]
```diff
@@ -477,6 +497,36 @@ int f9_test(void);
...
 
+#ifdef LTC_CHACHA20POLY1305_MODE
+
+typedef struct {
+   poly_state poly;
+   chacha_state chacha;
+   ulong64 aadlen;
+   ulong64 ctlen;
+   int aadflg;
+} chachapoly_state;
```
Comment (karel-m) [inline src/headers/tomcrypt_mac.h]: done

File: src/headers/tomcrypt_prng.h [hunk 9]
```diff
@@ -15,6 +15,15 @@ struct rc4_prng {
 };
 #endif
 
+#ifdef LTC_CHACHA
+struct chacha_prng {
```
Comment (karel-m) [inline src/headers/tomcrypt_prng.h]: done

File: src/headers/tomcrypt_prng.h [hunk 10]
```diff
@@ -55,6 +64,9 @@ typedef union Prng_state {
 #ifdef LTC_RC4
     struct rc4_prng       rc4;
 #endif
+#ifdef LTC_CHACHA
```
Comment (karel-m) [inline src/headers/tomcrypt_prng.h]: done

File: src/mac/poly1305/poly1305_file.c [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+   size_t x;
+   int err;
+
+   LTC_ARGCHK(fname  != NULL);
+   LTC_ARGCHK(key    != NULL);
+   LTC_ARGCHK(mac    != NULL);
+   LTC_ARGCHK(maclen != NULL);
+
+   if ((in = fopen(fname, "rb")) == NULL)                   { return CRYPT_FILE_NOTFOUND; }
+   if ((buf = XMALLOC(8196)) == NULL)                       { return CRYPT_MEM; }
```
Comment (karel-m) [inline src/mac/poly1305/poly1305_file.c]: Like
```C
#define FILE_READ_BUFSIZE 8192
```
somewhere in `tomcrypt_custom.h `?

File: src/prngs/chacha.c (hunk 3 above)
Comment (karel-m) [inline src/prngs/chacha.c]: A review would be great. 

Should I create a separate issue like `Call for reviewers - PRNG/ChaCha20` ?

File: src/headers/tomcrypt_custom.h (hunk 1 above)
Comment (karel-m) [inline src/headers/tomcrypt_custom.h]: done

File: src/mac/poly1305/poly1305_file.c (hunk 11 above)
Comment (karel-m) [inline src/mac/poly1305/poly1305_file.c]: done cd5e4e8

File: src/headers/tomcrypt_custom.h [hunk 12]
```diff
@@ -550,7 +550,10 @@
 
 #endif
 
-
+#ifndef LTC_NO_FILE
+   /* buffer size for reading from a file via fread(..) */
+   #define FILE_READ_BUFSIZE 8192
```
Comment (sjaeckel) [inline src/headers/tomcrypt_custom.h]:
```
#ifndef LTC_FILE_READ_BUFSIZE
#define LTC_FILE_READ_BUFSIZE 8192
#endif
```

Comment (karel-m) [inline src/headers/tomcrypt_custom.h]: done e57ca44

**Comment by karel-m** [comment]
BTW do we have a fixed file which I can use as an input in a test for `poly1305_file(..)`?

**Comment by sjaeckel** [comment]
> BTW do we have a fixed file which I can use as an input in a test for `poly1305_file(..)`?

we don't have tests for any of the `***_file()` API's... we can leave it like that I'd say... or use one of the testvector files?

**Comment by karel-m** [comment]
test vector files are sometimes changing (see sha3)

I was thinking about creating let's say `testprof/fixed-test-file.txt` with 10000 bytes of random text.

**Comment by karel-m** [comment]
You are perhaps right the *_file functions need a facelift anyway (to use the same read buffer style) which is worth a separate PR where we can add more tests as well.

**Comment by karel-m** [comment]
@sjaeckel can I merge this PR?

File: src/prngs/chacha.c [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+ *
+ * The library is free for all purposes without any express
+ * guarantee it works.
+ */
+
+#include "tomcrypt.h"
+
+#ifdef LTC_CHACHA20_PRNG
+
+const struct ltc_prng_descriptor chacha_prng_desc =
```
Comment (sjaeckel) [inline src/prngs/chacha.c]: I'd say we should also rename all these occurrences to `chacha20_prng***`

File: src/mac/poly1305/poly1305_file.c [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+   LTC_ARGCHK(key    != NULL);
+   LTC_ARGCHK(mac    != NULL);
+   LTC_ARGCHK(maclen != NULL);
+
+   if ((in = fopen(fname, "rb")) == NULL)                   { return CRYPT_FILE_NOTFOUND; }
+   if ((buf = XMALLOC(LTC_FILE_READ_BUFSIZE)) == NULL)      { return CRYPT_MEM; }
+   if ((err = poly1305_init(&st, key, keylen)) != CRYPT_OK) { goto LBL_ERR; }
+
+   do {
+      x = fread(buf, 1, sizeof(buf), in);
```
Comment (sjaeckel) [inline src/mac/poly1305/poly1305_file.c]: `sizeof(buf)` is either 4 or 8... :)

File: src/prngs/chacha.c (hunk 3 above)
Comment (sjaeckel) [inline src/prngs/chacha.c]: Sure, can you then please also link the existing implementation(s).

File: src/prngs/chacha.c (hunk 13 above)
Comment (karel-m) [inline src/prngs/chacha.c]: fixed ce37498

File: src/mac/poly1305/poly1305_file.c (hunk 14 above)
Comment (karel-m) [inline src/mac/poly1305/poly1305_file.c]: OMG, fixed 7b93f04

**Pull request merged**